File: pdi.cpp

// This is NWNC (No Warranty No Copyright) Software.
// astar.ai
// May 15, 2019

#include <GL/gl.h>
#include <opencv2/opencv.hpp>
#include "glwindow/scenewindow.hpp"

////////////////////////////////////////////////////////////////////////////////

int ndisp_bar = 1,   wsize_bar = 2, thr_bar = 0;
int ndisp_max = 2,   wsize_max = 3, thr_max = 130;
int ndisp_now = 32,  wsize_now = 9, thr_now = 190;

////////////////////////////////////////////////////////////////////////////////

void OnTrackNdisp(int, void*) {
  ndisp_now = 16 + 16 * ndisp_bar;
}

////////////////////////////////////////////////////////////////////////////////

void OnTrackWsize(int, void*) {
  wsize_now = 5 + 2 * wsize_bar;
}

////////////////////////////////////////////////////////////////////////////////

void OnTrackThreshold(int, void*) {
  thr_now = 190 + thr_bar;
}

////////////////////////////////////////////////////////////////////////////////

int     cap_cols, cap_rows, img_width;
cv::Mat T, Kl, Kr, Dl, Dr, xil, xir, Rl, Rr;
cv::Mat fmap[2], lmap[2][2], Kfe;

// CaliCam stereo is calibrated at 2560x960
// If you want to process image at 1280x480, set half_size to true.

bool half_size = true;

void LoadParameters(std::string file_name) {
  cv::FileStorage fs(file_name, cv::FileStorage::READ);
  if (!fs.isOpened()) {
    std::cout << "Failed to open ini parameters" << std::endl;
    exit(-1);
  }

  cv::Size cap_size;
  fs["cap_size" ] >> cap_size;
  fs["Kl"       ] >> Kl;
  fs["Dl"       ] >> Dl;
  fs["xil"      ] >> xil;
  fs["Rl"       ] >> Rl;
  fs["Kr"       ] >> Kr;
  fs["Dr"       ] >> Dr;
  fs["xir"      ] >> xir;
  fs["Rr"       ] >> Rr;
  fs["T"        ] >> T;
  fs.release();

  if (half_size) {
    cap_size  = cap_size / 2;
    Kl.row(0) = Kl.row(0) / 2.;
    Kl.row(1) = Kl.row(1) / 2.;
    Kr.row(0) = Kr.row(0) / 2.;
    Kr.row(1) = Kr.row(1) / 2.;
  }

  cap_cols  = cap_size.width;
  cap_rows  = cap_size.height;
  img_width = cap_size.width / 2;
}

////////////////////////////////////////////////////////////////////////////////

inline double MatRowMul(cv::Mat m, double x, double y, double z, int r) {
  return m.at<double>(r,0) * x + m.at<double>(r,1) * y + m.at<double>(r,2) * z;
}

////////////////////////////////////////////////////////////////////////////////

enum RectMode {
  RECT_PERSPECTIVE,
  RECT_FISHEYE,
  RECT_LONGLAT
};

void InitRectifyMap(cv::Mat K,
                    cv::Mat D,
                    cv::Mat R,
                    cv::Mat Knew,
                    double xi0,
                    double xi1,
                    cv::Size size,
                    RectMode mode,
                    cv::Mat& map1,
                    cv::Mat& map2) {
  map1.create(size, CV_32F);
  map2.create(size, CV_32F);

  double fx = K.at<double>(0,0);
  double fy = K.at<double>(1,1);
  double cx = K.at<double>(0,2);
  double cy = K.at<double>(1,2);
  double s  = K.at<double>(0,1);

  double k1 = D.at<double>(0,0);
  double k2 = D.at<double>(0,1);
  double p1 = D.at<double>(0,2);
  double p2 = D.at<double>(0,3);

  cv::Mat Ki  = Knew.inv();
  cv::Mat Ri  = R.inv();
  cv::Mat KRi = (Knew * R).inv();

  for (int r = 0; r < size.height; ++r) {
    for (int c = 0; c < size.width; ++c) {
      double xc = 0.;
      double yc = 0.;
      double zc = 0.;

      if (mode == RECT_PERSPECTIVE) {
        xc = MatRowMul(KRi, c, r, 1., 0);
        yc = MatRowMul(KRi, c, r, 1., 1);
        zc = MatRowMul(KRi, c, r, 1., 2);
      }

      if (mode == RECT_LONGLAT) {
        double tt = MatRowMul(Ki, c, r, 1., 0);
        double pp = MatRowMul(Ki, c, r, 1., 1);

        double xn = -cos(tt);
        double yn = -sin(tt) * cos(pp);
        double zn =  sin(tt) * sin(pp);

        xc = MatRowMul(Ri, xn, yn, zn, 0);
        yc = MatRowMul(Ri, xn, yn, zn, 1);
        zc = MatRowMul(Ri, xn, yn, zn, 2);
      }

      if (mode == RECT_FISHEYE) {
        double ee = MatRowMul(Ki, c, r, 1., 0);
        double ff = MatRowMul(Ki, c, r, 1., 1);

        double ef = ee * ee + ff * ff;
        double zz = (xi1 + sqrt(1. + (1. - xi1 * xi1) * ef)) / (ef + 1.);

        double xn = zz * ee;
        double yn = zz * ff;
        double zn = zz - xi1;

        xc = MatRowMul(Ri, xn, yn, zn, 0);
        yc = MatRowMul(Ri, xn, yn, zn, 1);
        zc = MatRowMul(Ri, xn, yn, zn, 2);
      }

      double rr = sqrt(xc * xc + yc * yc + zc * zc);
      double xs = xc / rr;
      double ys = yc / rr;
      double zs = zc / rr;

      double xu = xs / (zs + xi0);
      double yu = ys / (zs + xi0);

      double r2 = xu * xu + yu * yu;
      double r4 = r2 * r2;
      double xd = (1+k1*r2+k2*r4)*xu + 2*p1*xu*yu + p2*(r2+2*xu*xu);
      double yd = (1+k1*r2+k2*r4)*yu + 2*p2*xu*yu + p1*(r2+2*yu*yu);

      double u = fx * xd + s * yd + cx;
      double v = fy * yd + cy;

      map1.at<float>(r,c) = (float) u;
      map2.at<float>(r,c) = (float) v;
    }
  }
}

////////////////////////////////////////////////////////////////////////////////

int rect_cols = 640, rect_rows = 480;

void InitRectifyMap() {
  Kfe = cv::Mat::eye(3, 3, CV_64F);
  Kfe.at<double>(0,0) = Kl.at<double>(0,0) * 0.4;

  if (half_size)
    Kfe.at<double>(0,0) = Kl.at<double>(0,0) * 0.8;

  Kfe.at<double>(0,2) = rect_cols  / 2. - 0.5;
  Kfe.at<double>(1,1) = Kfe.at<double>(0,0);
  Kfe.at<double>(1,2) = rect_rows / 2. - 0.5;

  cv::Size img_size(rect_cols, rect_rows);
  cv::Mat  Kll = cv::Mat::eye(3, 3, CV_64F);
  Kll.at<double>(0,0) = img_size.width  / M_PI;
  Kll.at<double>(1,1) = img_size.height / M_PI;

  InitRectifyMap(Kl, Dl, Rl, Kfe, xil.at<double>(0,0), xil.at<double>(0,0),
                 img_size, RECT_FISHEYE, fmap[0], fmap[1]);
  InitRectifyMap(Kl, Dl, Rl, Kll, xil.at<double>(0,0), 0.,
                 img_size, RECT_LONGLAT, lmap[0][0], lmap[0][1]);
  InitRectifyMap(Kr, Dr, Rr, Kll, xir.at<double>(0,0), 0.,
                 img_size, RECT_LONGLAT, lmap[1][0], lmap[1][1]);
}

////////////////////////////////////////////////////////////////////////////////

void DisparityImage(const cv::Mat& recl, const cv::Mat& recr, cv::Mat& dispf) {
  cv::Mat disps;
  int N = ndisp_now, W = wsize_now, C = recl.channels();
  if (1) {
    cv::Ptr<cv::StereoSGBM> sgbm =
        cv::StereoSGBM::create(0, N, W, 8 * C * W * W, 32 * C * W * W);
    sgbm->compute(recl, recr, disps);
  } else {
    cv::Mat grayl, grayr;
    cv::cvtColor(recl, grayl, CV_BGR2GRAY);
    cv::cvtColor(recr, grayr, CV_BGR2GRAY);

    cv::Ptr<cv::StereoBM> sbm = cv::StereoBM::create(N, W);
    sbm->setPreFilterCap(31);
    sbm->setMinDisparity(0);
    sbm->setTextureThreshold(10);
    sbm->setUniquenessRatio(15);
    sbm->setSpeckleWindowSize(100);
    sbm->setSpeckleRange(32);
    sbm->setDisp12MaxDiff(1);
    sbm->compute(grayl, grayr, disps);
  }

  disps.convertTo(dispf, CV_32F, 1.f / 16.f);
}

////////////////////////////////////////////////////////////////////////////////

struct PCL {
  cv::Vec3f pts;
  cv::Vec3b clr;
};

void PointClouds(const cv::Mat& disp_img,
                 const cv::Mat& feim,
                 std::vector<PCL>& pcl_vec) {
  double fx = Kfe.at<double>(0,0);
  double fy = Kfe.at<double>(1,1);
  double cx = Kfe.at<double>(0,2);
  double cy = Kfe.at<double>(1,2);
  double bl = cv::norm(T);
  double xi = xil.at<double>(0,0);

  double w_pi = rect_cols  / M_PI;
  double h_pi = rect_rows / M_PI;
  double pi_w = M_PI / rect_cols;

  for (int r = 0; r < feim.rows; ++r) {
    for (int c = 0; c < feim.cols; ++c) {
      if (hypot(c - cx, r - cy) > thr_now)
        continue;

      double ee = (c - cx) / fx;
      double ff = (r - cy) / fy;
      double ef = ee * ee + ff * ff;
      double zz = (xi + sqrt(1. + (1. - xi*xi) * ef)) / (ef + 1.);

      if (isnan(zz))
        continue;

      double xs = zz * ee;
      double ys = zz * ff;
      double zs = zz - xi;

      double tt   = acos(-xs);
      double pp   = acos(-ys / hypot(ys, zs));
      float  disp = disp_img.at<float>((int) (pp * h_pi), (int) (tt * w_pi));

      if (disp <= 0.f)
        continue;

      double diff = pi_w * disp;
      double mgnt = bl * sin(tt - diff) / sin(diff);

      cv::Vec3b color = feim.at<cv::Vec3b>(r,c);

      PCL pcl;
      pcl.pts = cv::Vec3f(xs, ys, zs) * mgnt;
      pcl.clr = cv::Vec3b(color(2), color(1), color(0));
      pcl_vec.push_back(pcl);
    }
  }
}

////////////////////////////////////////////////////////////////////////////////

void DrawScene(const std::vector<PCL>& pcl_vec) {
  glBegin(GL_POINTS);

  for (uint i = 0; i < pcl_vec.size(); ++i) {
    PCL pcl = pcl_vec[i];
    glColor3ub(pcl.clr(0), pcl.clr(1), pcl.clr(2));
    glVertex3f(pcl.pts(0), pcl.pts(1), pcl.pts(2));
  }

  glEnd();
}

////////////////////////////////////////////////////////////////////////////////

bool live = false; // To run live mode, you need a CaliCam from www.astar.ai

int main(int argc, char** argv) {
  std::string param_name = argc >= 2 ? argv[1] : "./astar_calicam.yml";
  std::string image_name = argc == 3 ? argv[2] : "./wm_garden.jpg";

  LoadParameters(param_name);
  InitRectifyMap();

  glwindow::SceneWindow scene(960, 720, "Panorama 3D Scene");

  cv::VideoCapture vcapture;
  if (live) {
    vcapture.open(0);

    if (!vcapture.isOpened()) {
      std::cout << "Camera doesn't work" << std::endl;
      exit(-1);
    }

    vcapture.set(CV_CAP_PROP_FRAME_WIDTH,  cap_cols);
    vcapture.set(CV_CAP_PROP_FRAME_HEIGHT, cap_rows);
    vcapture.set(CV_CAP_PROP_FPS, 30);
  }

  std::string win_name = "Fisheye Image";
  cv::namedWindow(win_name);
  cv::createTrackbar("Num Disp:  16 + 16 *", win_name,
                     &ndisp_bar,  ndisp_max,   OnTrackNdisp);
  cv::createTrackbar("Blk   Size :     5  +  2 * ", win_name,
                     &wsize_bar,  wsize_max,  OnTrackWsize);
  cv::createTrackbar("Threshold :     190 + ", win_name,
                     &thr_bar,    thr_max,    OnTrackThreshold);

  cv::Mat raw_img = cv::imread(image_name, cv::IMREAD_COLOR);
  cv::Mat raw_imgl, raw_imgr, fe_img, ll_imgl, ll_imgr;

  if (half_size)
    cv::resize(raw_img, raw_img, cv::Size(), 0.5, 0.5);

  while (1) {
    if (live)
      vcapture >> raw_img;

    if (raw_img.total() == 0) {
      std::cout << "Image capture error" << std::endl;
      exit(-1);
    }

    raw_img(cv::Rect(        0, 0, img_width, cap_rows)).copyTo(raw_imgl);
    raw_img(cv::Rect(img_width, 0, img_width, cap_rows)).copyTo(raw_imgr);

    cv::remap(raw_imgl, fe_img,  fmap[0],    fmap[1],    cv::INTER_LINEAR);
    cv::remap(raw_imgl, ll_imgl, lmap[0][0], lmap[0][1], cv::INTER_LINEAR);
    cv::remap(raw_imgr, ll_imgr, lmap[1][0], lmap[1][1], cv::INTER_LINEAR);

    cv::Mat          disp_img;
    std::vector<PCL> pcl_vec;

    DisparityImage(ll_imgl, ll_imgr, disp_img);
    PointClouds(disp_img, fe_img, pcl_vec);

    if (scene.win.alive()) {
      if (scene.start_draw()) {
        DrawScene(pcl_vec);
        scene.finish_draw();
      }
    }

    imshow(win_name, fe_img);

    char key = cv::waitKey(1);
    if (key == 'q' || key == 'Q' || key == 27)
      break;
  }

  return 0;
}

////////////////////////////////////////////////////////////////////////////////

File: pdi.py

import sys
import cv2
import math
import numpy as np

#-------------------------------------------------------------------------------#

lmap = [[None, None], [None, None]]

cap_cols = None
cap_rows = None
img_width = None

width_now = 640
height_now = 640
thr_now = 0.75

ndisp_now = 32
wsize_now = 9

is_sgbm = True

margin = 15. * math.pi / 180.

#-------------------------------------------------------------------------------#

def load_parameters(param_file):
  global Kl, Dl, Rl, xil, Kr, Dr, Rr, xir, T
  global cap_cols, cap_rows, img_width

  fs = cv2.FileStorage(param_file, cv2.FILE_STORAGE_READ)

  cap_size_node = fs.getNode("cap_size")
  cap_cols = int(cap_size_node.at(0).real())
  cap_rows = int(cap_size_node.at(1).real())
  img_width = cap_cols

  Kl = fs.getNode("Kl").mat()
  Dl = fs.getNode("Dl").mat()
  Rl = fs.getNode("Rl").mat()
  xil = fs.getNode("xil").mat()

  Kr = fs.getNode("Kr").mat()
  Dr = fs.getNode("Dr").mat()
  Rr = fs.getNode("Rr").mat()
  xir = fs.getNode("xir").mat()

  T = fs.getNode("T").mat()

  img_width = cap_cols / 2

#-------------------------------------------------------------------------------#

def init_undistort_rectify_map(k, d, r, knew, xi0, size, mode):
  fx = k[0, 0]
  fy = k[1, 1]
  cx = k[0, 2]
  cy = k[1, 2]
  s  = k[0, 1]

  k1 = d[0, 0]
  k2 = d[0, 1]
  p1 = d[0, 2]
  p2 = d[0, 3]

  ki = np.linalg.inv(knew)
  ri = np.linalg.inv(r)
  kri = np.linalg.inv(np.matmul(knew, r))

  rows = size[0]
  cols = size[1]

  mapx = np.zeros((rows, cols), dtype = np.float32)
  mapy = np.zeros((rows, cols), dtype = np.float32)

  for r in range(rows):
    for c in range(cols):
      xc = 0.0
      yc = 0.0
      zc = 0.0

      if mode == 'kRectPerspective':
        cr1 = np.array([c, r, 1.])
        xc = np.dot(kri[0, :], cr1)
        yc = np.dot(kri[1, :], cr1)
        zc = np.dot(kri[2, :], cr1)

      if mode == 'kRectLonglat':
        cr1 = np.array([c, r, 1.])
        tt = np.dot(ki[0, :], cr1)
        pp = np.dot(ki[1, :], cr1)

        xn = -math.cos(tt)
        yn = -math.sin(tt) * math.cos(pp)
        zn =  math.sin(tt) * math.sin(pp)

        cr1 = np.array([xn, yn, zn])
        xc = np.dot(ri[0, :], cr1)
        yc = np.dot(ri[1, :], cr1)
        zc = np.dot(ri[2, :], cr1)

      if mode == 'kRectFisheye':
        cr1 = np.array([c, r, 1.])
        ee = np.dot(ki[0, :], cr1)
        ff = np.dot(ki[1, :], cr1)
        zz = 2. / (ee * ee + ff * ff + 1.)

        xn = zz * ee
        yn = zz * ff
        zn = zz - 1.

        cr1 = np.array([xn, yn, zn])
        xc = np.dot(ri[0, :], cr1)
        yc = np.dot(ri[1, :], cr1)
        zc = np.dot(ri[2, :], cr1)


      if mode == 'kRectCylindrical':
        cr1 = np.array([c, r, 1.])
        tt = np.dot(ki[0, :], cr1)
        pp = np.dot(ki[1, :], cr1) + margin

        xc = -math.sin(pp) * math.cos(tt)
        yc = -math.cos(pp)
        zc =  math.sin(pp) * math.sin(tt)


      if zc < 0.0:
        mapx[r, c] = np.float32(-1.)
        mapy[r, c] = np.float32(-1.)

        continue

      rr = math.sqrt(xc * xc + yc * yc + zc * zc)
      xs = xc / rr
      ys = yc / rr
      zs = zc / rr

      xu = xs / (zs + xi0)
      yu = ys / (zs + xi0)

      r2 = xu * xu + yu * yu
      r4 = r2 * r2
      xd = (1 + k1 * r2 + k2 * r4) * xu + 2 * p1 * xu * yu + p2 * (r2 + 2 * xu * xu)
      yd = (1 + k1 * r2 + k2 * r4) * yu + 2 * p2 * xu * yu + p1 * (r2 + 2 * yu * yu)

      u = fx * xd + s * yd + cx
      v = fy * yd + cy

      mapx[r, c] = np.float32(u)
      mapy[r, c] = np.float32(v)

  return mapx, mapy

#-------------------------------------------------------------------------------#

def init_rectify_map():
  global mode, fmap, lmap, Kfisheye

  ### longlat
  kRectLonglat = np.identity(3, dtype = np.float64)
  kRectLonglat[0, 0] = (width_now - 1.) / math.pi
  kRectLonglat[1, 1] = (height_now - 1.) / math.pi

  img_size = [height_now, width_now]

  print("Initialize left longlat rectify map")
  lmap[0][0], lmap[0][1] = init_undistort_rectify_map(
      Kl, Dl, Rl, kRectLonglat, xil, img_size, 'kRectLonglat')

  print("Initialize right longlat rectify map")
  lmap[1][0], lmap[1][1] = init_undistort_rectify_map(
      Kr, Dr, Rr, kRectLonglat, xir, img_size, 'kRectLonglat')

#-------------------------------------------------------------------------------#

def disparity_image(rect_imgl, rect_imgr):
  if is_sgbm == True:
    stereo = cv2.StereoSGBM_create(0, ndisp_now, wsize_now, 
                                   3 * 8 * wsize_now * wsize_now, 
                                   3 * 32 * wsize_now * wsize_now)
  else:
    stereo = cv2.StereoBM_create(ndisp_now, wsize_now)

  disparity = stereo.compute(rect_imgl, rect_imgr)
  disparity = disparity / 16.

  return disparity

#-------------------------------------------------------------------------------#

# point cloud generation!
# Credit: Thanks Yawen Lu <yl4280@rit.edu> for providing ply saving routine

ply_header = '''ply
format ascii 1.0
element vertex %(vert_num)d
property float x
property float y
property float z
property uchar red
property uchar green
property uchar blue
end_header
'''

def write_ply(fn, disp_img, rect_imgl_longlat):
   bl = cv2.norm(T)
   pi_w = math.pi / (width_now - 1.)

   colors = np.zeros_like(rect_imgl_longlat, dtype='f')
   points = np.zeros_like(rect_imgl_longlat, dtype='f')
   colors_img = cv2.cvtColor(rect_imgl_longlat, cv2.COLOR_BGR2RGB)
   rows1 = colors_img.shape[0]
   cols1 = colors_img.shape[1]

   for r in range(rows1):
     for c in range(cols1):
        disp = disp_img[r, c]
        if (disp <= 0.):
           continue

        tt = (c / (cols1 - 1.) - 0.5) * math.pi
        pp = (r / (rows1 - 1.) - 0.5) * math.pi
        
        cx = math.sin(tt)
        cy = math.cos(tt) * math.sin(pp)
        cz = math.cos(tt) * math.cos(pp)
        
        cr = math.hypot(math.hypot(cx, cy), cz)
        ct = math.acos(cz / cr)
        
        if ct > (math.pi / 2.) * thr_now:
           continue
        
        diff = pi_w * disp
        mgnt = bl * math.sin(c * pi_w - diff) / math.sin(diff)
        
        points[r, c] = np.array([cx, cy, cz], dtype='f') * mgnt
        colors[r, c] =  colors_img[r, c]

   points = points.reshape(-1, 3)
   colors = colors.reshape(-1, 3)
   verts = np.hstack([points, colors])
   with open(fn, 'wb') as f:
        f.write((ply_header % dict(vert_num=len(verts))).encode('utf-8'))
        np.savetxt(f, verts, fmt='%f %f %f %d %d %d ')

#-------------------------------------------------------------------------------#

def main():
  param_file = "astar_calicam.yml"
  image_name = "wm_garden.jpg"

  if len(sys.argv) == 2:
    param_file = sys.argv[1]

  if len(sys.argv) == 3:
    param_file = sys.argv[1]
    image_name = sys.argv[2]

  load_parameters(param_file)
  init_rectify_map()

  raw_img = cv2.imread(image_name, 1)

  raw_imgl = raw_img[:, : int(img_width)]
  raw_imgr = raw_img[:, int(img_width) : int(img_width * 2)]

  rect_imgl_longlat = cv2.remap(raw_imgl, lmap[0][0], lmap[0][1], cv2.INTER_LINEAR)
  rect_imgr_longlat = cv2.remap(raw_imgr, lmap[1][0], lmap[1][1], cv2.INTER_LINEAR)

  dim = (int(img_width / 2), int(cap_rows / 2))
  small_img = cv2.resize(raw_imgl, dim, interpolation = cv2.INTER_NEAREST)
  cv2.imshow("Raw Image", small_img)

  disp_img = disparity_image(rect_imgl_longlat, rect_imgr_longlat)

  ### point cloud generation
  out_fn = 'pdi_point_cloud.ply'

  print('Save %s' % out_fn)
  write_ply(out_fn, disp_img, rect_imgl_longlat)

  print('Press \'q\' to finish')
  key = cv2.waitKey(0)

  if key & 0xFF == ord('q') or key  == 27:
    return

#-------------------------------------------------------------------------------#

if __name__ == "__main__":
  main()

#-------------------------------------------------------------------------------#
